File: tests/test_leg_kinematics.cc
#include <vector>
#include <math.h>
#include <iomanip>
#include <stdexcept>
#include <iostream>
using namespace std;

double l1 = 50, l2 = 20, l3 = 120, l4 = 155;

vector<double> leg_kinematics(vector<double> point) {
  double x = point[0], y = point[1], z = point[2];
  double _l1 = l1, _l2 = l2,_l3 = l3,_l4 = l4;
  double f = 0;
  double theta3 = 0;

  if(isnan(sqrt(pow(x, 2) + pow(y, 2) - pow(_l1, 2)))) {
    cout << "NAN in the leg kinematics with x: " << x << ", y: " << y << " and l1: " << _l1 << "\n";
    f = _l1;
  } else {
    f = sqrt(pow(x, 2) + pow(y, 2) - pow(l1, 2));
  }

  double g = f - _l2;
  double h = sqrt(pow(g, 2) + pow(z, 2));
  cout << h << endl;

  double theta1 =- atan2(y, x) - atan2(f, -_l1);
  cout << theta1 << endl;

  double d = ((pow(h, 2) - pow(_l3, 2) - pow(_l4, 2)) / (2 * _l3 * _l4));
  cout << d << endl;

  if (isnan(acos(d))) {
    cout << "NAN in the leg kinematics with x: " << x << ", y: " << y << " and d: " << d << "\n";
    theta3 = 0;
  } else {
    theta3 = acos(d);
  }
  cout << theta3 << endl;
  double theta2 = atan2(z, g) - atan2(_l4 * sin(theta3), _l3 + _l4 * cos(theta3));

  return vector<double> {theta1, theta2, theta3};
}

int main() {
  vector<double> a = {1200, 1200, 1200};
  vector<double> b = leg_kinematics(a);

  for(size_t i = 0; i < b.size(); i++) { 
    cout << b[i] << " ";
  }
  cout << "\n";
}

File: tests/test_leg_kinematics.py
import numpy as np
from math import *
import random

l1=50
l2=20
l3=120
l4=155

def legIK(point):
        (x,y,z)=(point[0],point[1],point[2])
        (_l1,_l2,_l3,_l4)=(l1,l2,l3,l4)
        try:
            F=sqrt(x**2+y**2-_l1**2)
        except ValueError:
            print("Error in legIK with x {} y {} and _l1 {}".format(x,y,_l1))
            F=_l1
        G=F-_l2
        H=sqrt(G**2+z**2)
        print(H)
        theta1=-atan2(y,x)-atan2(F,-_l1)
        print(theta1)
        D=(H**2-_l3**2-_l4**2)/(2*_l3*_l4)
        print(D)
        try:
            theta3=acos(D)
        except ValueError:
            print("Error in legIK with x {} y {} and D {}".format(x,y,D))
            theta3=0
        theta2=atan2(z,G)-atan2(_l4*sin(theta3),_l3+_l4*cos(theta3))

        return(theta1,theta2,theta3)

a = np.array([1200, 1200, 1200])
b = legIK(a)

print(b)